# BEE 4750 Lab 3: Linear Programming with JuMP

**Name**: Cella Schnabel

**ID**: rms428

> **Due Date**
>
> Wednesday, 10/16/24, 9:00pm

## Setup

The following code should go at the top of most Julia scripts; it will
load the local package environment and install any needed packages. You
will see this often and shouldn’t need to touch it.

In [1]:
import Pkg
Pkg.activate(".")
Pkg.instantiate()

  Activating project at `~/Desktop/BEE4750/labs/lab-3-rms428`
   Installed HiGHS_jll ──────────── v1.7.2+0
   Installed CodecBzip2 ─────────── v0.8.3
   Installed StaticArraysCore ───── v1.4.3
   Installed CommonSubexpressions ─ v0.3.0
   Installed BenchmarkTools ─────── v1.5.0
   Installed MutableArithmetics ─── v1.4.6
   Installed DiffRules ──────────── v1.15.1
   Installed DiffResults ────────── v1.1.0
   Installed HiGHS ──────────────── v1.9.2
   Installed ForwardDiff ────────── v0.10.36
   Installed MathOptInterface ───── v1.31.0
   Installed JuMP ───────────────── v1.22.2
Precompiling project...
  ✓ StaticArraysCore
  ✓ CodecBzip2
  ✓ CommonSubexpressions
  ✓ DiffRules
  ✓ HiGHS_jll
  ✓ DiffResults
  ✓ BenchmarkTools
  ✓ Qt6ShaderTools_jll
  ✓ FFMPEG_jll
  ✓ Qt6Declarative_jll
  ✓ FFMPEG
  ✓ GR_jll
  ✓ MutableArithmetics
  ✓ ForwardDiff
  ✓ GR
  ✓ Plots
  ✓ Plots → UnitfulExt
  ✓ MathOptInterface
  ✓ HiGHS
  ✓ JuMP
  20 dependencies successfully precompiled in 54 seconds. 150 alr

In [2]:
using JuMP # optimization modeling syntax
using HiGHS # optimization solver

## Overview

In this lab, you will write and solve a resource allocation example
using `JuMP.jl`. `JuMP.jl` provides an intuitive syntax for writing,
solving, and querying optimization problems.

`JuMP` requires the loading of a solver. \[Each supported solver works
for certain classes of problems, and some are open source while others
require a commercial license\]. We will use the `HiGHS` solver, which is
open source and works for linear, mixed integer linear, and quadratic
programs.

In this lab we will walk through the steps involved in coding a linear
program in HiGHS, solving it, and querying the solution.

## Exercise (3 points)

Your task is to decide how much lumber to produce to maximize profit
from wood sales. You can purchase wood from a managed forest, which
consists of spruce (320,000 bf) and fir (720,000 bf). Spruce costs
$\$0.12$ per bf to purchase and fir costs $\$0.08$ per bf.

At the lumber mill, wood can be turned into plywood of various grades
(see <a href="#tbl-inputs" class="quarto-xref">Table 1</a> for how much
wood of each type is required for and the revenue from each grade). Any
excess wood is sent to be recycled into particle board, which yields no
revenue for the mill.

| Plywood Grade | Inputs (bf/bf plywood) | Revenue (\$/1000 bf) |
|:-------------:|:----------------------:|:--------------------:|
|       1       |   0.5 (S) + 1.5 (F)    |         400          |
|       2       |   1.0 (S) + 2.0 (F)    |         520          |
|       3       |   1.5 (S) + 2.0 (F)    |         700          |

Table 1: Wood inputs and revenue by plywood grade. S refers to spruce
inputs, F fir inputs.

First, we need to identify our decision variables. While there are
several options, we will use $G_i$, the amount of each grade the mill
produces (in \$/1000 bf).

Using these decision variables, formulate a linear program to maximize
the profit of the mill subject to the supply constraints on spruce and
fir.

> **JuMP Syntax**
>
> The core pieces of setting up a `JuMP` model involve specifying the
> model and adding variables, the objective, and constraints. At the
> most simple level, this syntax looks like this:
>
> ``` julia
> m = Model(HiGHS.Optimizer)
> @variable(m, lb <= x <= ub) # if you do not have upper or lower bounds, you can drop those accordingly
> @variable(m, lb <= y <= ub)
> @objective(m, Max, 100x + 250y) # replace Max with Min depending on the problem
> @constraint(m, label, 6x + 12y <= 80) # replace "label" with some meaningful string you would like to use later to query shadow prices, or drop it
> ```
>
> You can add more constraints or more variables as needed.

> **Using Array Syntax**
>
> You can set up multiple variables or constraints at once using array
> syntax. For example, the following are equivalent:
>
> ``` julia
> @variable(m, G1 >= 0)
> @variable(m, G2 >= 0)
> @variable(m, G3 >= 0)
> ```
>
> and
>
> ``` julia
> @variable(m, G[1:3] >= 0)
> ```
>
> You can also set up multiple constraints using arrays of coefficients
> and/or bounds. For example:
>
> ``` julia
> I = 1:3
> d = [0; 3; 5]
> @constraint(m, demand[i in I], G[i] >= d[i])
> ```

`JuMP` is finicky about changing objects and constraints, so I recommend
setting up all of the model syntax in one notebook cell, which is what
we will do here.

**Objectives and Constraints**\
The objective is to maximize profit using $G_i$, the amount of each grade, $i$, the mill produces (in \$/1000 bf). \
$Profit = Revenue - Cost$ \
I found the following three profit formulas based on which Grade, $G_i$\
$G_1$: $(400 - 120*0.5 - 80*1.5)G_1$\
$G_2$: $(520 - 120*1 - 80*2)G_2$\
$G_3$: $(400 - 120*1.5 - 80*2)G_3$\
I combine these to maximize the following equation:\
$max$: $220*G_1+240*G_2+360*G_3$

The constraints of the model are based on the amount of spruce and firs available in the forest.\
Spruce Constraint: $500*G_1+1000*G_2+1500*G_3 \leq 320000 $\
Fir Contraint: $1500*G_1+2000*G_2+2000*G_3 \leq 720000 $

In [17]:
forest_model = Model(HiGHS.Optimizer) # initialize model object
@variable(forest_model, G[1:3] >= 0) # non-negativity constraints
# uncomment the following lines and add the objective and constraints as needed for the model
@objective(forest_model, Max, 220*G[1]+240*G[2]+360*G[3])
@constraint(forest_model, forestspruce, 500*G[1]+1000*G[2]+1500*G[3] <= 320000)
@constraint(forest_model, forestfir, 1500*G[1]+2000*G[2]+2000*G[3] <= 720000)
print(forest_model) # this outputs a nicely formatted summary of the model so you can check your specification

Max 220 G[1] + 240 G[2] + 360 G[3]
Subject to
 forestspruce : 500 G[1] + 1000 G[2] + 1500 G[3] ≤ 320000
 forestfir : 1500 G[1] + 2000 G[2] + 2000 G[3] ≤ 720000
 G[1] ≥ 0
 G[2] ≥ 0
 G[3] ≥ 0


Next, to optimize, use the `optimize!()` function:

In [18]:
optimize!(forest_model)

Running HiGHS 1.7.2 (git hash: 5ce7a2753): Copyright (c) 2024 HiGHS under MIT licence terms
Coefficient ranges:
  Matrix [5e+02, 2e+03]
  Cost   [2e+02, 4e+02]
  Bound  [0e+00, 0e+00]
  RHS    [3e+05, 7e+05]
Presolving model
2 rows, 3 cols, 6 nonzeros  0s
2 rows, 3 cols, 6 nonzeros  0s
Presolve : Reductions: rows 2(-0); columns 3(-0); elements 6(-0) - Not reduced
Problem not reduced by presolve: solving the LP
Using EKK dual simplex solver - serial
  Iteration        Objective     Infeasibilities num(sum)
          0    -6.2499928695e-01 Ph1: 2(6.5918); Du: 3(0.624999) 0s
          2     1.1200000000e+05 Pr: 0(0) 0s
Model   status      : Optimal
Simplex   iterations: 2
Objective value     :  1.1200000000e+05
HiGHS run time      :          0.01


You should get confirmation that a solution was found; if one was not,
there’s a chance something was wrong with your model formulation.

To find the values of the decision variables, use `value()` (which can
be broadcasted over variable arrays):

In [19]:
@show value.(G);

value.(G) = [352.0, 0.0, 96.0]


**Solution**\
From the `optimize!` and `value` calls of the model, I found that the optimized solution orders 352,000 bf of $G_1$, 0 bf of $G_2$, and 96,000 bf of $G_3$

Similarly, `objective_value()` finds the optimal value of the objective:

In [20]:
@show objective_value(forest_model);

objective_value(forest_model) = 112000.0


**Objective Value** Is your total profit, or the maximum value of your objective (our objective was profit). I found the total profit to be 112000 dollars.

Finally, we can find the dual values of the constraints with
`shadow_price()`. Do this for the constraints in your model using the
block below.

In [21]:
@show shadow_price(forestspruce);
@show shadow_price(forestfir);

shadow_price(forestspruce) = 0.08000000000000002
shadow_price(forestfir) = 0.12


**Shadow Price**\
Because both spruce and fir are non-zero values, if you change the constraint on either of the values, it changes how much profit you make. If the shadow price was zero for either constraint, than changing the constraints on the variables wouldn't change your profit.

`JuMP` also lets you evaluate other expressions that you might be
interested in based on the solutions. For example, you can use the
following block to calculate the total amount of plywood the mill would
produce under the optimal solution:

In [22]:
@expression(forest_model, total_plywood, sum(G))
@show value.(total_plywood);

value.(total_plywood) = 448.0


**Total Plywood**
Total plywood is how many grades you sell in total, or $G_1+G_2+G_3 = 448$ (which checks, as $352+96 = 448$)

## References

Put any consulted sources here, including classmates you worked with/who
helped you.

Tis and I worked on this together. He helped me with the units and the explanation of the shadow price.